### PR TITLE
Change contact_type  to contact_types

### PR DESCRIPTION
**Pull request by jtu0**
### TODO BEFORE MERGE
 - [x] _Fix the places in the codebase assume a single value for `contact_type`_
 - [x] _Add back in the validation!_
 - [x] _Fix the database seeds_
 - [x] _Get all specs passing_

**Nice to have but not necessary before merge**
- [ ] _Fix the labels not being clickable in the new case contact from (perhaps by creating a partial or a form helper for bootstrap checkboxes)_
- [ ] _Add specs for case_contact_report model_
- [x] volunteer dashboard view: use a decorator instead of ugly chaining in the view

**Zoe should add as followup issues**
- [ ] case contact report has new headers - open new issue if needs rename
----

### Description
Moved `contact_type` from string to array of strings.

Resolves https://github.com/rubyforgood/casa/issues/119

### Type of change
New feature (non-breaking change which adds functionality)

*Warning* This change assumes there is no data that can be lost.

### How will this affect user permissions?

user permissions will not be impacted by this change - "no impact"

### How Has This Been Tested?
As a Volunteer, create a new Case Contact.
When all other parts of the form are completed validly,
Selecting 1 or more `Contact Types` should successfully submit the form.
Selecting no `Contact Type` should be a form validation error.

### Screenshots
*FIXME ADD THESE*

### CODE REVIEW

**Comment by compwron** [comment]
Yay it is so cool that y'all are working on this :)

**Comment by zspencer** [comment]
@colinsoleim  - That makes sense for dealing with invalid but present values; but my understanding is we want an empty set to also be an invalid value.

My gut says there is a missing test for when updating the `contact_types` to be an empty set; and the solution will be to add a validation that's a bit grumpier than the built in `presence` validation; as it appears the presence validation and the `array` type don't refuse empty arrays.

I'm not going to have time to poke at this til next Wednesday; so feel free to carry it on if you want! But also happy to carry it across the finish line :)

**Comment by zspencer** [comment]
Also @compwron  - Yea, @jtu0 was like "Now that we have so much free time... why not start streaming again?"

And I said no for like 3 days and then said "OK fine" and now here we are :)

**Comment by jtu0** [comment]
Oh no, did *I* bring this on the both of us??!!??

This whole time I'd been wondering what had possessed @zspencer to want to start live streaming or me to agree and now I'm realizing this was all my fault :laughing:

**Comment by zspencer** [comment]
... :-o ...

Well, now I'm having fun so let's give it a quarter and re-assess :D :D :D

**Comment by colinsoleim** [comment]
> @colinsoleim - That makes sense for dealing with invalid but present values; but my understanding is we want an empty set to also be an invalid value.
> 
> My gut says there is a missing test for when updating the `contact_types` to be an empty set; and the solution will be to add a validation that's a bit grumpier than the built in `presence` validation; as it appears the presence validation and the `array` type don't refuse empty arrays.
> 
> I'm not going to have time to poke at this til next Wednesday; so feel free to carry it on if you want! But also happy to carry it across the finish line :)

I just double checked because I wasn't sure how the presence validator worked with array fields, but it looks like it does give an error for an empty array. I checked the API docs, and the presence validator calls `Object#blank?` which rejects empty strings, nil values, and empty arrays. Do you think that will be enough?

Screen Shot 2020-04-22 at 8 56 54 PM

**Comment by zspencer** [comment]
Hmm, this leaves me even more befuddled; as the test that was sending an empty array was "creating" a CaseContact which is why we pended it out.

Ah well, I'll dig in next Wednesday.

File: app/models/case_contact.rb [hunk 1]
```diff
@@ -21,10 +19,40 @@ class CaseContact < ApplicationRecord
   ].freeze
 
   CONTACT_MEDIUMS = %w[in-person text/email video voice-only letter].freeze
-  enum contact_type: CONTACT_TYPES.zip(CONTACT_TYPES).to_h
+  # enum contact_type: CONTACT_TYPES.zip(CONTACT_TYPES).to_h
```
Comment (compwron) [inline app/models/case_contact.rb:22]: In general, please delete code rather than leaving it commented out :)

File: app/models/case_contact.rb [hunk 2]
```diff
@@ -21,10 +19,40 @@ class CaseContact < ApplicationRecord
...
+  validate :contact_types_included
+
+  def contact_types_included
+    contact_types&.each do |contact_type|
+      errors.add(:contact_types, :invalid) unless CONTACT_TYPES.include? contact_type
+    end
+  end
+
+
+  # @return [Array<String>]
```
Comment (compwron) [inline app/models/case_contact.rb:34]: Cool, types! How is this used? Is it an IDE plugin?

**Comment by colinsoleim** [comment]
This looks great to me!

File: app/models/case_contact.rb [hunk 3]
```diff
@@ -21,10 +19,40 @@ class CaseContact < ApplicationRecord
...
+  def humanized_contact_types
+    contact_types&.map { |ct| ct.humanize.titleize }
```
Comment (compwron) [inline app/models/case_contact.rb:36]: I think that because of how rails associations work, this should never be null, but it could be an empty array.

File: app/models/case_contact.rb [hunk 4]
```diff
@@ -34,7 +62,7 @@ def humanized_type
 #
 #  id               :bigint           not null, primary key
 #  contact_made     :boolean          default(FALSE)
-#  contact_type     :string           not null
+#  contact_types    :string           is an Array
```
Comment (compwron) [inline app/models/case_contact.rb:65]: COOL

File: app/models/case_contact.rb [hunk 5]
```diff
@@ -46,8 +74,9 @@ def humanized_type
 #
 # Indexes
 #
-#  index_case_contacts_on_casa_case_id  (casa_case_id)
-#  index_case_contacts_on_creator_id    (creator_id)
+#  index_case_contacts_on_casa_case_id   (casa_case_id)
+#  index_case_contacts_on_contact_types  (contact_types) USING gin
```
Comment (compwron) [inline app/models/case_contact.rb:78]: gin?

File: app/models/case_contact_report.rb [hunk 6]
```diff
@@ -12,12 +12,14 @@ def self.to_csv
...
 
   def self.report_headers
-    headers = %w[ internal_contact_number duration contact_type contact_made
-                  contact_medium occurred_at added_to_system_at casa_case_number
-                  volunteer_email volunteer_name supervisor_name]
+    headers = %w[internal_contact_number duration]
+
+    headers.concat(CaseContact::CONTACT_TYPES.map { |t| "contact_type: #{t}" })
+
+    headers.concat(%w[ contact_made
```
Comment (compwron) [inline app/models/case_contact_report.rb:19]: This formatting is weird- one item on a line, then longer lines below. It might be better to make it one item per line if they don't all fit on one line.

File: app/views/case_contacts/index.html.erb [hunk 7]
```diff
@@ -19,7 +19,7 @@
       <tr>
         <td><%= case_contact.creator_id %></td>
         <td><%= case_contact.casa_case_id %></td>
-        <td><%= case_contact.contact_type %></td>
+        <td><%= case_contact.humanized_contact_types&.join(', ') %></td>
```
Comment (compwron) [inline app/views/case_contacts/index.html.erb:22]: This would be a good candidate to put in a decorator

File: db/migrate/20200422180727_replace_contact_type_with_contact_types_on_case_contact.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class ReplaceContactTypeWithContactTypesOnCaseContact < ActiveRecord::Migration[6.0]
+  def change
+
+    # NOTE: This is a destructive migration that we would normally avoid
+    #       if we were working on production data, but because there is
+    #       no production data we are comfortable being destructive and
+    #       losting whatever data is in the `contact_type` column.
+    remove_column :case_contacts, :contact_type
+    add_column :case_contacts, :contact_types, :string, array: true
+    add_index :case_contacts, :contact_types, using: :gin
```
Comment (compwron) [inline db/migrate/20200422180727_replace_contact_type_with_contact_types_on_case_contact.rb:10]: aha https://hashrocket.com/blog/posts/exploring-postgres-gin-index

File: spec/features/volunteer_adds_a_case_contact_spec.rb [hunk 9]
```diff
@@ -10,15 +10,19 @@
...
 
-    click_on 'Submit'
+    expect  do
+      click_on 'Submit'
+    end.to change(CaseContact, :count).by(1)
 
     expect(CaseContact.first.casa_case_id).to eq volunteer_casa_case_one.id
-    expect(CaseContact.first.contact_type).to eq 'school'
+    expect(CaseContact.first.contact_types).to include 'school'
+    expect(CaseContact.first.contact_types).to include 'therapist'
```
Comment (compwron) [inline spec/features/volunteer_adds_a_case_contact_spec.rb:25]: :)

File: spec/requests/case_contacts_spec.rb [hunk 10]
```diff
@@ -90,6 +90,11 @@
     end
 
     context 'with invalid parameters' do
+      # Right now, there is no way for users to edit CaseContacts
+      # And this test is failing for a reason we don't understand.
+      #
+      # We have pended it out because we are not worried about
```
Comment (compwron) [inline spec/requests/case_contacts_spec.rb:96]: This test appears to be no longer pended, so we should remove the out of date comment

**Comment by compwron** [comment]
This looks great and safe to merge!
After it gets to staging, I'll re-run db:seed to generate us some case contact types :)

File: db/migrate/20200422180727_replace_contact_type_with_contact_types_on_case_contact.rb (hunk 8 above)
Comment (compwron) [reply db/migrate/20200422180727_replace_contact_type_with_contact_types_on_case_contact.rb:10]: It isn't really valid to have no contact type on a contact- is there a way to add that requirement at a db level?

**Comment by compwron** [comment]
Y'all are so great! I also love the twitch stream 💖

File: app/models/case_contact.rb (hunk 2 above)
Comment (jtu0) [reply app/models/case_contact.rb]: Yep for vscode.

File: app/models/case_contact.rb (hunk 5 above)
Comment (zspencer) [reply app/models/case_contact.rb:71]: Updated with a comment on the migration!

File: app/models/case_contact_report.rb (hunk 6 above)
Comment (zspencer) [reply app/models/case_contact_report.rb]: Good point!

File: app/views/case_contacts/index.html.erb (hunk 7 above)
Comment (zspencer) [reply app/views/case_contacts/index.html.erb]: Done!

File: app/models/case_contact.rb (hunk 3 above)
Comment (jtu0) [reply app/models/case_contact.rb]: WE THOUGHT SO TOO but it turned out to be a necessary `&` :woman_shrugging:

File: db/migrate/20200422180727_replace_contact_type_with_contact_types_on_case_contact.rb (hunk 8 above)
Comment (zspencer) [reply db/migrate/20200422180727_replace_contact_type_with_contact_types_on_case_contact.rb:16]: We definitely could add a not null constraint; but not this week. Mind making an issue and tagging us in it?

File: spec/requests/case_contacts_spec.rb (hunk 10 above)
Comment (zspencer) [reply spec/requests/case_contacts_spec.rb]: Good catch! Removed!
